codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java

```diff
@@ -780,53 +807,59 @@ private LastHttpContent readTrailingHeaders(ByteBuf buffer) {
     protected abstract HttpMessage createMessage(String[] initialLine) throws Exception;
     protected abstract HttpMessage createInvalidMessage();
 
-    private static int getChunkSize(String hex) {
-        hex = hex.trim();
-        for (int i = 0; i < hex.length(); i ++) {
-            char c = hex.charAt(i);
-            if (c == ';' || Character.isWhitespace(c) || Character.isISOControl(c)) {
-                hex = hex.substring(0, i);
-                break;
+    private static int getChunkSize(byte[] hex, int length) {
+        // hex shouldn't contain both ISO chars nor spaces
+        int result = 0;
+        for (int i = 0; i < length; i++) {
+            final int digit = StringUtil.decodeHexNibble(hex[i]);
+            if (digit == -1) {
+                throw new NumberFormatException();
             }
+            result *= 16;
+            result += digit;
         }
-
-        return Integer.parseInt(hex, 16);
+        return result;
     }
 
-    private static String[] splitInitialLine(AppendableCharSequence sb) {
+    private static String[] splitInitialLine(ByteBuf sb) {
         int aStart;
         int aEnd;
         int bStart;
         int bEnd;
         int cStart;
         int cEnd;
 
-        aStart = findNonSPLenient(sb, 0);
-        aEnd = findSPLenient(sb, aStart);
+        final byte[] sbContent = sb.array();
 
-        bStart = findNonSPLenient(sb, aEnd);
-        bEnd = findSPLenient(sb, bStart);
+        final int length = sb.readableBytes();
 
-        cStart = findNonSPLenient(sb, bEnd);
-        cEnd = findEndOfString(sb);
+        aStart = findNonSPLenient(sbContent, 0, length);
+        aEnd = findSPLenient(sbContent, aStart, length);
+
+        bStart = findNonSPLenient(sbContent, aEnd, length);
+        bEnd = findSPLenient(sbContent, bStart, length);
+
+        cStart = findNonSPLenient(sbContent, bEnd, length);
+        cEnd = findEndOfString(sbContent, length);
 
         return new String[] {
-                sb.subStringUnsafe(aStart, aEnd),
-                sb.subStringUnsafe(bStart, bEnd),
-                cStart < cEnd? sb.subStringUnsafe(cStart, cEnd) : "" };
+                sb.toString(aStart, aEnd - aStart, CharsetUtil.US_ASCII),
+                sb.toString(bStart, bEnd - bStart, CharsetUtil.US_ASCII),
+                cStart < cEnd ? sb.toString(cStart, cEnd - cStart, CharsetUtil.US_ASCII) : ""};
     }
 
-    private void splitHeader(AppendableCharSequence sb) {
-        final int length = sb.length();
+    private void splitHeader(ByteBuf sb) {
+        final byte[] sbContent = sb.array();
+        final int length = sb.readableBytes();
         int nameStart;
         int nameEnd;
         int colonEnd;
         int valueStart;
         int valueEnd;
 
-        nameStart = findNonWhitespace(sb, 0);
+        nameStart = findNonWhitespace(sbContent, 0, length);
```

Comment:
Same comment as above 
```suggestion
        nameStart = findNonWhitespace(sbContent, sb.readerIndex() + sb.arrayOffset(), length);
```